### DISCUSSION TITLE
Strip headers/cookies from request before rewrite

### DISCUSSION THREAD

**Discussion by fjeldstad** [vercel/next.js#36987] [Help]
I would like to use a middleware as a simple proxy for an external API, and it works fine. _But_ it forwards the request exactly as-is, which is expected but unfortunate as I don't want it to include cookies and potentially other sensitive information in the call to the external service.

Is there a way to modify the request headers in the middleware before forwarding it via `NextResponse.rewrite`? I have tested `req.headers.delete('cookie')` but no luck (and no error).

Example `/pages/api/_middleware.js`:
```javascript
import { NextResponse } from 'next/server';

export default async function middleware(request) {
  const baseUrl = 'https://postman-echo.com';
  const url = request.nextUrl.clone();
  const path = url.pathname.replace(/^\/api/, '');
  const rewrittenUrl = `${baseUrl}/${request.method.toLowerCase()}${path}${
    url.search
  }`;
  // TODO: Remove `cookie` header (and potentially more) from the request
  return NextResponse.rewrite(rewrittenUrl);
}

```

**Comment by icyJoseph** [comment]
I had tried this before, but for internal routes, and it didn't really work out back then.

In this case, would it work out to make the call to the external API, and then send the response back to the origin? I guess it'd still have the same effect of a rewrite, letting you access the external API through your application. After all, fetch is available in the edge runtime so it should work for just about any REST call.

**Reply by fjeldstad** [reply]
I suppose that would work. For now I'm using a catch-all API route that internally uses `http2-proxy` which offers async hooks for modifying request and response. For my specific use case there is no requirement for running the rewrite/proxy at the edge, although that would of course be nice for the general case. I think the most important aspect is that it is as transparent and lightweight as possible.

**Reply by 4lejandrito** [reply]
Hi! I have the same need. I'm using rewrites to [proxy an analytics script through the same domain in next-plausible](https://github.com/4lejandrito/next-plausible#proxy-the-analytics-script). It works fine but the cookies get sent to the analytics server which is not desired for security and performance reasons (see [this issue](https://github.com/4lejandrito/next-plausible/issues/67)).

I've found a workaround using middleware, but the problem is that next-plausible is a library that automatically configures the rewrites and cannot set the middleware transparently. It'd be great if rewrites could have an additional property `credentials: 'omit'` to tell next not to forward cookies on rewrites:

```js
{
  source: '/path',
  destination: 'http://example.com',
  credentials: 'omit',
},
```

What do you think?
